keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1250 @@
+
+<!--
+Things to keep in mind include: additional in-memory state, additional
+non-trivial computations, excessive access to disks (including increased log
+volume), significant amount of data sent and/or received over network, etc.
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+<!--
+Focus not just on happy cases, but primarily on more pathological cases
+(e.g. probes taking a minute instead of milliseconds, failed pods consuming resources, etc.).
+If any of the resources can be exhausted, how this is mitigated with the existing limits
+(e.g. pods per node) or new limits added by this KEP?
+
+Are there any tests that were run/should be run to understand performance characteristics better
+and validate the declared limits?
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Future Improvements
```

Comment:
Any plans for statefulsets?

Reply:
I can mention it there. I would prefer not to do it in the first iteration. I think we need an option on the `NodeMaintenance` or/and on the `StatefulSet` to support this. The main problem is that it could be a disruptive action (e.g. StatefulSet with 1 pod) and we are trying to design the API to give more control to the application.

Reply:
**After the first alpha**, a higher-level controller that owns a StatefulSet could manage scale out for the StatefulSet to account for the upcoming (or currently happening) maintenance.

There is no one-size-fits-all answer. For some apps, the right fix might be to add another temporary replica. For another app, it is to quiesce, snapshot and then delete the running Pod so that a replacement can come up in the same zone, avoiding the node that is tainted due to maintenance. In another situation, maybe there are always exactly 3&times;<i>n</i> replicas, and we use in-place vertical scaling&hellip;

Does this sound familiar? It's the [operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/), right? My advice is to offer the NodeMaintenance API and teach higher-level abstractions - outside Kubernetes' core - to care about it. We can even offer an official CRD for older clusters.

Reply:
FYI The use cases the users brought up to me are mostly statefulset related so far.

Reply:
@soltysh @sftim @dchen1107 

ReplicaSet, Deployment, StatefulSet are discussed in the https://github.com/kubernetes/enhancements/pull/4565 KEP now.
I have added empty sections to this KEP for DaemonSet and Static pods. And will discuss it there.